### PR TITLE
current draft for primal storm pets

### PR DESCRIPTION
**Pull request by cyriun**
detection for the "Elementally Imbued" buff seems to work but its still giving attempts whether the buff is present or not. For example, add a random npc id that doesn't have the correct buff to the primal storm pets in the database and it will give an attempt.

### CODE REVIEW

**Comment by Duckwhale** [comment]
Assuming that the logic is correct (haven't tested it ingame), the code would behave as follows:

* Rarity calls `IsAttemptAllowed` to determine whether an attempt should be added
* Your code iterates through all active (helpful) auras on the target unit, if any exists
* It tries to match every one of them against the required aura ID and name
* Since there is no match, it proceeds executing the rest of the function
* This will usually return `true`, leading Rarity to add an attempt for the item

In other words, you forgot `return` (or `return false`) in the case of `requireAuraNotFound`. Does that sound about right?

**Comment by cyriun** [comment]
gotcha, thanks i believe i do know what you mean, will get back to testing.

**Comment by cyriun** [comment]
ok after adding return false it does seem to disallow attempts when the aura isnt found. So now i believe all that has to be done is re-adding all the relevant npc ids

**Comment by Duckwhale** [comment]
Just FYI: I haven't forgotten about this at all, but I probably won't have time to review/test before the weekend. Thanks, though!

File: DB/Pets/Dragonflight.lua [hunk 1]
```diff
@@ -771,47 +766,42 @@ local dragonflightPets = {
...
 			203756,
 			203757,
+			203754,
+			203755,
 		},
 		itemId = 200255,
 		spellId = 389363,
 		chance = 1000,
 		creatureId = 189112,
+		requiredAura = { [CONSTANTS.AURAS.ELEMENTALLY_IMBUED] = true },
```
[the remark this answers is not in the document]
Comment (cyriun) [reply DB/Pets/Dragonflight.lua:804]: i think the `requiredAura` on brul currently is useless because it doesnt work inside the `OnMassiveGainsOpened` function

File: Core.lua [hunk 2]
```diff
@@ -673,6 +673,202 @@ function R:IsAttemptAllowed(item)
...
+		201557,
+		201558,
+		201559,
+		203744,
+	}
+	local requiredAuraFound = false
+
+	if item.requiredAura then
+		local unit = "target"
+		local npcid = tonumber((select(6, strsplit("-", UnitGUID("target")))) or 0)
```
[the remark this answers is not in the document]
Comment (cyriun) [reply Core.lua:838]: yep `R:GetNPCIDFromGUID(guid)` ?

File: Core.lua [hunk 3]
```diff
@@ -673,6 +673,202 @@ function R:IsAttemptAllowed(item)
 		return false
 	end
 
+	local buffedNPCS = {
```
[the remark this answers is not in the document]
Comment (cyriun) [reply Core.lua:676]: yea, cant remember if that was a mistake or intentional

[the remark this answers is not in the document]
Comment (cyriun) [reply Core.lua:676]: nvm, it has to be in the db also because of npc method right?

[the remark this answers is not in the document]
Comment (Duckwhale) [reply Core.lua:676]: The main idea would be de-duplication, as otherwise the lists need to be kept in sync.

File: Core.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (Duckwhale) [reply Core.lua:838]: Yeah, that's the one.

File: DB/Pets/Dragonflight.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (Duckwhale) [reply DB/Pets/Dragonflight.lua:804]: What does that mean? I'm a bit confused by the different implementations. Ideally, we want just one (that's consistent) :D

[the remark this answers is not in the document]
Comment (cyriun) [reply DB/Pets/Dragonflight.lua:804]: well for instance passing item obviously works in the `R:IsAttemptAllowed` function in core.lua but in the `OnMassiveGainsOpened` function passing item didnt seem to work

[the remark this answers is not in the document]
Comment (Duckwhale) [reply DB/Pets/Dragonflight.lua:804]: If you have an `item` table, you can pass it wherever you need. But beware `:` (`self`), the implicit first argument.

[the remark this answers is not in the document]
Comment (cyriun) [reply DB/Pets/Dragonflight.lua:804]: not too sure what you mean by this

[the remark this answers is not in the document]
Comment (Duckwhale) [reply DB/Pets/Dragonflight.lua:804]: You can pass the `item` table to any function (as an argument), after you've obtained it from the saved variables.

**Comment by Duckwhale** [comment]
What's the state of this PR? I've rebased it to resolve the conflicts, but there seems to be at least one error:

> 11x Rarity/Core.lua:845: bad argument #2 to 'strsplit' (string expected, got nil)

Should I spend more time getting this to work or shelve it for now? I only took a brief look, so maybe I missed something.

**Comment by cyriun** [comment]
> What's the state of this PR? I've rebased it to resolve the conflicts, but there seems to be at least one error:
> 
> > 11x Rarity/Core.lua:845: bad argument #2 to 'strsplit' (string expected, got nil)
> 
> Should I spend more time getting this to work or shelve it for now? I only took a brief look, so maybe I missed something.

ill have to resub soon and take a look but i believe it was something to do with npcids duplicated?

**Comment by Duckwhale** [comment]
No need to resub just for this! But it seemed like the NPC ID was `nil`, so I guess something's a bit funky here still.

I'll mark this as a draft for now. Hopefully, I'll find some time to review things in detail before too long.

**Comment by cyriun** [comment]
> No need to resub just for this! But it seemed like the NPC ID was `nil`, so I guess something's a bit funky here still.
> 
> I'll mark this as a draft for now. Hopefully, I'll find some time to review things in detail before too long.

i was planning to resub soon anyways, sub dropped off over christmas and new years and i havent been back on yet but im sure i will be soon

**Comment by Duckwhale** [comment]
After spending more time on the problem, I think there are some flaws with the approach of checking for auras:

* As already noted, the "persistent" (elementals/caster) NPCs don't have any auras - this complicates the handling code, a lot
* The other NPCs are empowered by an element based on the event, and should only drop the pet of the "right" element
* We also can't check NPC IDs in the function that checks for allowed attempts, since it only provides the item DB entry
* Lastly, even if we could - by the time the NPC is looted the aura will no longer be active (since the NPC is dead, it has no auras)

So if you don't mind), I'd close this and proceed later with an approach based on Area POIs (using [these IDs](https://old.wow.tools/dbc/?dbc=areapoi&build=10.0.5.47660#page=1&search=Elemental%20Storm), see WIP branch [here](https://github.com/WowRarity/Rarity/tree/655-primal-storm-detection)).

**Comment by Duckwhale** [comment]
Closing in favor of #720. Thank you for your effort!
